# INTRODUCTION TO PYTHON

### CONTENT
1. Introduction
2. Variables and data structures 
3. Conditional and iteractive structures
4. Functions 
5. Introduction to NumPy 
6. Introduction to Pandas
7. Exercises

## 1. Introduction

<img src="Figures/Jupyter.png" alt="Drawing" style="width: 100px;"/>

Jupyter notebooks is an environment (IDE) that allows to interleave text content (to be called 'Markdown' cell) with Python code ('Code' cell) that can be executed cell by cell as shown in the following examples.

The basic commands to start using Jupyter Notebooks are:

<img src="Figures/menu-vEng.png" alt="Drawing" style="width: 850px;"/>

* `Shift+Enter`: Execute cell and select next cell.
* `Alt+Enter`: Execute cell and insert a new cell.
* `ESC`: to start 'command' mode. Once in this mode:
    * `H` list of commands
    * `A` insert cell below
    * `B` inster cell above
    * `D,D` D twice to delete cell
    * `Y` change cell type to 'Code'
    * `M` change cell type to 'Markdown'
    * `Enter` go back to 'edit' mode.

#### Execute the cell below using any of the possible commands

In [1]:
3 + 4 + 9

16

In [2]:
# Python code example
# Change values a and b, and observe the result when executing the cell 

a=87
b=34
c=a+b

print('The result of adding {} and {} is {}.'.format(a,b,c))

The result of adding 87 and 34 is 121.


### 1.1. Variables and data structures

Variables in Python are created once they are defined, that is, when a value is assigned to the variable for the first time (using the <code> = </code> operator). For this reason, variables (and their type) do not need to be declared in Python. 

The main types of variables that can be found are:
* **Numeric**
    * **Integer:** Positive and negative natural numbers.
    * **Float:** Real numbers.
    * **Complex:** Complex numbers. The imaginary part is multiplied by `j` to represent the root of `-1` Números complejos.
* **Boolean**
    These are variables that can take True or False as a value, represented in Python by `True` and `False` using uppercase for the initial letter.
* **Data structures (data sequences)**.
    Ordered list of values of the same or different type. In Python they exist:
    * **String:** text strings
    * **List:** ordered sets of elements
    * **Tuple:** ordered and immutable sets of elements
    * **Dictionary:** sets of elements characterized by an identifier and a value  

There are some general rules for variable names:
+ Only letters, numbers and "underscores" (_) are used.
+ NO spaces are used anywhere in the name
+ The name cannot start with a number
+ Uppercase letters are treated as different from lowercase letters (i.e., Python is case-sensitive)

In [3]:
# The name can't start with a number
#3a = 10

Whenever a variable is created or modified, Python interprets what type of variable it is. When a new value is assigned to an existing object (*dynamic typing*), the previous values of that object are cleared from the computer's memory. The first value of **a** is the real number 13.0, but after a new assignment:

In [4]:
# Dynamic typing
a = 13.0
a = 'b'
print(a)

b


The ``type()`` function displays the TYPE of the data we introduce:

In [5]:
x = True
print(type(x))

name = 'Vinicius'
print(type(name))

age = 32
print(type(age))

money = 5.5
print(type(money))

list1 = ['Vinicius', 32, 5.5]
print(type(list1))

<class 'bool'>
<class 'str'>
<class 'int'>
<class 'float'>
<class 'list'>


We can print the elements in the list or any variables using the command `print` and `.format` or simply summing strings:

In [6]:
print('My name is {}, I have {} years old and {} euros.'.format(list1[0], list1[1], list1[2]))

#or

print('My name is {}, I have {} years old and {} euros.'.format(name, age, money))

#or

print('My name is '+ name + ', I have ' + str(age) + ' years old and ' + str(money) + ' euros')

My name is Vinicius, I have 32 years old and 5.5 euros.
My name is Vinicius, I have 32 years old and 5.5 euros.
My name is Vinicius, I have 32 years old and 5.5 euros


### 1.2 Types and basic operators

The basic operators in Python are:

+ Addition (also for strings, tuples or lists): <code>a + b</code>
+ Subtraction: <code>a - b</code>
+ Multiplication (also for strings, tuples and lists): <code>a * b</code>
+ Division: <code>a / b</code>
+ Integer division: <code>a // b</code>
+ Remainder: <code>a % b</code>
+ Exponential: <code>a ** b</code>
+ Assignment: <code>=</code>, <code>-=</code>, <code>+=</code>,<code>/=</code>,<code>*=</code>, <code>%=</code>, <code>//=</code>, <code>**=</code>
+ Boolean comparisons: <code>==</code>, <code>!=</code>, <code><</code>,<code>></code>,<code><=</code>, <code>>=</code>
+ Boolean operations: <code>and</code>, <code>or</code>, <code>not</code>
+ Operations to check belongingness: <code>in</code>, <code>not in</code>
+ Operations to identify objects: <code>is</code>, <code>is not</code>
<!-- + Bitwise operators (or, xor, and, complement): <code>|</code>, <code>^</code>, <code>&</code>, <code>~</code> -->
<!-- + Left and right bit shift: <code><<</code>, <code>>></code> -->

In [7]:
# Boolean or logic operators

a = 7
b = 2

print(a>b and b<a)

print(a==7 and b==5)

print(a==7 or b==5)

True
False
True


In [8]:
# Operators to check belongingness

my_list = [1, 3, 2, 7, 9, 8, 6]
print(4 in my_list)
print(8 in my_list)

False
True


In [9]:
# Operators to identify objects

a = 7
b = 2  
c = 7
print(a is c)
print (a is not b)
print (a is not c)

True
True
False


### 1.3 Python as a calculator

Python has a concise notation for arithmetic that closely resembles the traditional way of writing operations.

In [10]:
a = 3+2
b= 3.5 * -8
c = 10/6
print(a, b, c, 10./6.)

5 -28.0 1.6666666666666667 1.6666666666666667


For more complex operations, we need to use other packages such as `math` and `numpy`.

In [11]:
import numpy as np   # only executed if the library hasn't been imported previously
print(np.pi)
print(np.sin(1.8))
print(np.arccos(0.8))

3.141592653589793
0.9738476308781951
0.6435011087932843


## 2 Data structures

### 2.1 Strings or text strings

Text in Python is treated as a list of characters. Thus, the methodology that applies to the treatment of lists can be applied to text strings.

In [12]:
print("This is 'an example' of how to use quotation marks inside a text that is written between double quotation marks")
print('Here is "another example" but this time inverted')

This is 'an example' of how to use quotation marks inside a text that is written between double quotation marks
Here is "another example" but this time inverted


The '+' operator can be used to join character strings:

In [13]:
a = 'He'
b = 'llo'
c = a+b+'!'
print(c)

Hello!


To access only a part of a list or, in this case, of a *string*, the technique called **slicing** is used. The format to follow to access only a part of the string is as follows:

### <center>variable_name[first_element_index:final_index:step]</center>

Where the operation of the indexes can be seen in the following picture:

![Slicing1](Figures/slicing1.png)


In [14]:
#Example
a = 'abcdef'

print('Everything: {}'.format(a[:]))
print('First value: {}'.format(a[0]))
print('Second to end: {}'.format(a[2:]))
print('First 3: {}'.format(a[:3]))
print('Second to Fourth: {}'.format(a[2:4]))
print('All odd positions: {}'.format(a[::2]))
print('All even positions: {}'.format(a[1::2]))

Everything: abcdef
First value: a
Second to end: cdef
First 3: abc
Second to Fourth: cd
All odd positions: ace
All even positions: bdf


In the case of *strings*, there are default methods in Python that can be useful. For example the following functions:
- len: returns the length of the text.
- str: converts a number into a string.
- reverse: returns the text backwards.

We can also use string-specific methods:
- lower: converts all text to lowercase.
- upper: converts all text to uppercase.
- capitalize: capitalizes the first letter of the string.
- split(delimiter): returns a list of substrings separated by the specified delimiter (see examples below).


In [15]:
text='We are going to test the Strings Functions and Methods!'
print(len(text))
print(text.lower())
print(text.split(' '))

# Return the 4th word of the text in all capital letters
print(text[7:12].upper())

55
we are going to test the strings functions and methods!
['We', 'are', 'going', 'to', 'test', 'the', 'Strings', 'Functions', 'and', 'Methods!']
GOING


### 2.2 Lists

Lists are a Python type used to group objects of other types (even lists!). For example a list of numbers, of floats, of strings, of other lists, and so on. 

Lists, like strings, can use the *slicing* methodology.


In [16]:
# Create list
my_list = [0,5,'6',34,'Hello']
my_list

[0, 5, '6', 34, 'Hello']

In [17]:
# Access to an element of a list
a = my_list[2]
a

'6'

In [18]:
# Change an element of a list
my_list[-1]='Bye'
my_list

[0, 5, '6', 34, 'Bye']

In [19]:
# We can have lists of lists
my_list=[45,[3,5,6],'word',34.6,['a','b','c']]
print(my_list)
print(my_list[1])
print(my_list[1][2])
my_list[4][2]

[45, [3, 5, 6], 'word', 34.6, ['a', 'b', 'c']]
[3, 5, 6]
6


'c'

For lists there are also default Python functions and methods that can be useful. For example the following functions:
- len: returns the length of the list.
- list: create list

To use a function, call function_name(variable_name).

We can also use list-specific methods:
- append: add element to the end of the list.
- insert: add an element in a particular position.
- pop: remove the last element of the list.
- remove: remove a particular element.
- sort: sort the list

To use a method, call variable_name.method_name.

In [20]:
my_list=['first','hello','test','bye','animals','master']
print(my_list)

my_list.append('last')
print(my_list)

my_list.insert(1,'second')
print(my_list)

my_list.pop()
print(my_list)

my_list.remove('test')
print(my_list)

my_list.sort()
print(my_list)

['first', 'hello', 'test', 'bye', 'animals', 'master']
['first', 'hello', 'test', 'bye', 'animals', 'master', 'last']
['first', 'second', 'hello', 'test', 'bye', 'animals', 'master', 'last']
['first', 'second', 'hello', 'test', 'bye', 'animals', 'master']
['first', 'second', 'hello', 'bye', 'animals', 'master']
['animals', 'bye', 'first', 'hello', 'master', 'second']


### 2.3 Dictionaries

A dictionary is another way of grouping data where an element can be accessed through its *key*. For example:

In [21]:
this_installation = {
    'Power': 10,
    'Type': 'Solar',
    'Construction_year':2012
}

In [22]:
this_installation['Type']

'Solar'

Another example could be:

In [23]:
dict = {"Installation_1": "Solar", "Installation_2": "Wind", "Installation_3": "Wind", "Installation_4": "Hidraulyc",
        "Installation_5": "Solar",}
print(dict)

{'Installation_1': 'Solar', 'Installation_2': 'Wind', 'Installation_3': 'Wind', 'Installation_4': 'Hidraulyc', 'Installation_5': 'Solar'}


To access, modify or add an item, it can be done in a similar way to the lists:

In [24]:
# Access
dict["Installation_1"]

'Solar'

In [25]:
# Modify
print(dict)
dict["Installation_1"]="Wind"
print(dict)

{'Installation_1': 'Solar', 'Installation_2': 'Wind', 'Installation_3': 'Wind', 'Installation_4': 'Hidraulyc', 'Installation_5': 'Solar'}
{'Installation_1': 'Wind', 'Installation_2': 'Wind', 'Installation_3': 'Wind', 'Installation_4': 'Hidraulyc', 'Installation_5': 'Solar'}


In [26]:
# Add
print(dict)
dict["Installation_6"]="Solar"
print(dict)

{'Installation_1': 'Wind', 'Installation_2': 'Wind', 'Installation_3': 'Wind', 'Installation_4': 'Hidraulyc', 'Installation_5': 'Solar'}
{'Installation_1': 'Wind', 'Installation_2': 'Wind', 'Installation_3': 'Wind', 'Installation_4': 'Hidraulyc', 'Installation_5': 'Solar', 'Installation_6': 'Solar'}


Dictionaries can contain all types (int, floats, lists, other dictionaries, etc.).

In [27]:
dict = {"Installation_1": ["Solar", '10 MW'], "Installation_2": ["Wind", '15 MW'], "Installation_3": ["Solar", '18 MW'],
        "Installation_4": ["Hydraulic", '12 MW'], "Installation_5": ["Solar", '23 MW'], "Installation_6": ["Wind", '8 MW'],
        "Installation_7": ["Solar", '11 MW'], "Installation_8": ["Hydraulic", '11 MW']}
dict

{'Installation_1': ['Solar', '10 MW'],
 'Installation_2': ['Wind', '15 MW'],
 'Installation_3': ['Solar', '18 MW'],
 'Installation_4': ['Hydraulic', '12 MW'],
 'Installation_5': ['Solar', '23 MW'],
 'Installation_6': ['Wind', '8 MW'],
 'Installation_7': ['Solar', '11 MW'],
 'Installation_8': ['Hydraulic', '11 MW']}

In [28]:
dict['Installation_3'][1]

'18 MW'

In [29]:
# Changing power of installation 4 to 15 MW modifying the dictionary!

dict['Installation_4'][1]='15 MW'
print(dict['Installation_4'])

['Hydraulic', '15 MW']


### 2.4 Tuples

Tuples are lists **that cannot be modified.**

Their syntax is the same as lists, but to write a tuple you use parentheses and not square brackets.

In [30]:
mysupertuple = ('Solar', 'Wind', 'Hydraulic', 'Gas', 'Nuclear', 'NoMoreCoal')
print(type(mysupertuple), mysupertuple[4:])

<class 'tuple'> ('Nuclear', 'NoMoreCoal')


In [31]:
#CANNOT BE MODIFIED!
#mysupertuple[0]='PV'


## 3. Conditional and iteractive structures

### 3.1 Conditional control structures (if)

Conditional control structures allow you to execute one part of the code or another depending on the evaluation of one or several boolean conditions of YES or NO (**<code>TRUE</code>** o **<code>FALSE</code>**).

In Python, conditional control structures are defined by the words **<code>if</code>**, **<code>elif</code>** y **<code>else</code>**.

- **<code>if CONDICION:</code>** if the conditional expression is met, the following code block is executed.
- **<code>elif CONDICION:</code>** otherwise, if this conditional expression is met, this other code block is executed.
- **<code>else:</code>** otherwise, this code block is executed without evaluating any condition.


The condition is usually evaluated by the following relational operators <code> <, <=, ==, >=, >, != </code>.

The part of code inside an **<code>if</code>** cannot be empty, but if for some reason an if statement must be created without content, the **<code>pass</code>** statement can be used to prevent an error from occurring.

In [32]:
# Example

celsius = 0
fahrenheit = 9.0 /5.0 * celsius + 32

print("Temperature in Fahrenheit is", fahrenheit)

if fahrenheit > 90:
    print("It is hot")
elif fahrenheit < 30:
    print("It is cold")
else: print("It is neither hot nor cold")

Temperature in Fahrenheit is 32.0
It is neither hot nor cold


The conditional structure can also be expressed in a single line as follows, depending on whether one, two or more conditions are involved:

In [33]:
#Example

a = 330
b = 330

# with one only condition:
if a > b: print("a is larger than b")
    
# with two conditions:
print("a is larger") if a > b else print("a is not larger")

# with 3 conditions:
print("A") if a > b else print("=") if a == b else print("B")

a is not larger
=


####  <font color='red'>Try it yourself!</font> 

<font color='red'>Create a code to identify whether the variable **<code>number</code>** is a positive, negative, or 0 value. Store the result in a variable called **<code>sign</code>**. </font> 

In [34]:
number = 2

###############################
if  number > 0:
    sign = '+'
elif number < 0:
    sign = '-'
else:
    sign = '0'

print(sign)

+


### 3.2 Iterative control structures (for & while)

Iterative control structures (also called loops) allow the same code to be executed repeatedly as long as a condition is met.

There are two types of iterative control structures:

* For
* While

The Python ``for`` statement iterates over the elements of any sequence (a list or a string for example), in the order they appear in the sequence.

The ``while`` loop allows multiple iterations to be performed based on the result of a logical expression that can result in a ``True`` or ``False`` value.

In [35]:
# Example of for

for i in range(0,10,1):
    print(i)

0
1
2
3
4
5
6
7
8
9


In [36]:
# Example of while

i=0
while i < 10 :
    print(i)
    i= i+1

0
1
2
3
4
5
6
7
8
9


In addition, loops with ``for`` or ``while`` can be combined with conditional ``if`` structures to create more complex structures.

With the ``break`` statement we can stop the loop before it has gone through all elements.

With the ``continue`` statement we can stop the current iteration of the loop (in both ``for`` and ``while``) and continue with the next one.

####  <font color='red'>Try it yourself!</font> 

<font color='red'>Create a code that calculates the average value of a list of **<code>numbers</code>** using a `for` loop.

In [37]:
# Exercise: calculating the average value of a list with a for

numbers=[2,2,2,2]

#####################

suma = 0
for n in numbers:
    suma = suma + n

average = suma / len(numbers)

print(numbers, average)

###########################
average = sum(numbers) / len(numbers)
print(numbers, "average =", average)

[2, 2, 2, 2] 2.0
[2, 2, 2, 2] average = 2.0


### 3.3 Lists (or dictionary) comprehensions

List comprehensions are a way of fitting a ``for`` loop, an ``if`` statement, and an assignment all on a single line.

A list comprehension consists of the following parts:

+ An input sequence
+ A variable representing members of the input sequence
+ An optional expression
+ An output expression that produces elements of the output list from members of the input sequence that satisfy the predicate

In [38]:
# Example
num = [1, 4, -5, 10, -7, 2, 3, -1]
squared = []
for i in num:
     if i > 0:
        squared.append(i**2)
print(type(squared), squared)

# This for can be rewritten as follows:
# In an iteration, it first checks if the if is TRUE and then performs the calculation
squared = [x**2 for x in num if x > 0] 
print(type(squared), squared)

<class 'list'> [1, 16, 100, 4, 9]
<class 'list'> [1, 16, 100, 4, 9]


This compact way of expressing loops can also be used to iterate within lists. Below are different ways to join the values of two lists using zip or nested fors.

In [39]:
list1 = ['a','b','c','d']
list2 = [1,2,3,4]

# Create a list with nested fors:
print ([(i,j) for i in list1 for j in list2])

# Create a list using zip:
print ([(i,j) for i,j in zip(list1,list2)])

# Cereate a dictionary using zip:
print ({i:j for i,j in zip(list1,list2)})

[('a', 1), ('a', 2), ('a', 3), ('a', 4), ('b', 1), ('b', 2), ('b', 3), ('b', 4), ('c', 1), ('c', 2), ('c', 3), ('c', 4), ('d', 1), ('d', 2), ('d', 3), ('d', 4)]
[('a', 1), ('b', 2), ('c', 3), ('d', 4)]
{'a': 1, 'b': 2, 'c': 3, 'd': 4}


## 4. Functions
A function is a block of code with an associated name, which receives zero or more arguments as input, follows a sequence of statements where the desired operations are executed and returns a value and/or performs a task. This code block can be called multiple times as needed.

In Python there are a number of functions built into the language by default, but you can also create user-defined functions to use in your own programs. 

The complete list of default built-in functions can be found at:
www.w3schools.com/python/python_ref_functions.asp

In [40]:
# Example of an integrated function in Python

# The input() function allows us to assign a value entered by the user to a variable

a = input()

print('The variable introduced has been ' + a)

The variable introduced has been 3


To create a function in Python, you must use the word ``def`` to define the function. The syntax for a function definition in Python is:

<code>*def* function_name(input_parameter_list):
    function_statements
    *return* [expression_to_return]</code>

Where:
* function_name: is the name of the function.
* input_parameter_list: is the list of parameters that a function can receive, separated by comma.
* function_statements: is the block of statements in Python source code that perform a given operation.
* expression_to_return: is the expression or variable that returns the return statement.

In [41]:
# Example: function to set the first value of a list to 0

def change_first_element(list):
    list[0]=0 #it returns none!

numbers=[1,2,3,4]
change_first_element(numbers)
print(numbers)

[0, 2, 3, 4]


#### Example: greatest common denominator (GCD)

The common denominator of two positive integers $a$ and $b$ is the greatest common divisor between $a$ and $b$ . The Euclidean algorithm is an iterative method for computing the greatest common denominator of two integers. In pseudocode it would be:

+ If $a<b$, change $a$ and $b$.
+ Divide $a$ by $b$ and obtain the residue, $r$. 
+ If $r=0$, the value $b$ is GCD of $a$ and $b$.
+ If $r \neq 0$, iterate again replacing $a$ by $b$ and replacing $b$ by $r$. 

In [42]:
def GCD(a,b):
    r = 1
    while r != 0:
        if a<b:
            c=a
            a=b
            b=c
        r = a%b 
        if r == 0:
            return b
        else:
            a = b
            b = r

GCD(100,16)

4

####  <font color='red'>Try it yourself!</font> 

<font color='red'>Create a function to calculate the factorial of an integer. The factorial of a non-negative integer $n$ is the product of all positive integers less than or equal to $n$.</font> 

In [43]:
#########################################

def fact_1(n):
    result = 1
    while n > 1:
        result *= n
        n -= 1
    return result

print(fact_1(5))

120


### 4.1 Arguments and parameters of functions

When defining a function, the values that are received are called parameters, while during the call, the values that are sent are called arguments.

By default, a function must be called with the correct number of arguments. If the function expects 2 arguments, you must call the function with 2 arguments.

#### Arguments per position

When arguments are sent to a function, by default they are received in order in the defined parameters, known as arguments by position. In the example above (GCD), argument 100 is position 0, which corresponds to the parameter of function a, while argument 16 is position 1, which corresponds to the parameter of function b.

#### Arguments by name

However, it is possible to skip the order of the parameters if you indicate during the call what value each parameter has from its name.

In [44]:
GCD(b=100,a=16)


4

#### Default parameters

To avoid error messages when a function is not called with the required arguments, you can define default parameters within a function.

In [45]:
def GCD(a=None,b=None):
    if a == None or b == None:
        print ("Error, you must send two numbers to the function.")
        return
    r = 1
    while r != 0:
        if a<b:
            c=a
            a=b
            b=c
        r = a%b 
        if r == 0:
            return b
        else:
            a = b
            b = r

GCD()

Error, you must send two numbers to the function.


#### ndeterminate parameters ``*args`` y ``**kwargs``

Sometimes it is not known in advance how many items need to be sent to a function.

If you do not know how many arguments per item will be passed to a function, you add a * before the parameter name in the function definition. By convention, the parameter name ``*args`` is usually used. In this way, the function will receive a tuple of arguments and can access the items accordingly.

Similarly, if it is not known how many arguments by name will be passed to a function, two ** are added before the parameter name in the function definition. By convention, the parameter name ``**kwargs`` is usually used. In this way, the function will receive a dictionary of arguments and can access the elements accordingly.

In [46]:
def super_function(*args,**kwargs):
    total = 0
    for arg in args:
        total += arg
    print ("sum => ", total)
    for kwarg in kwargs:
        print (kwarg, "=>", kwargs[kwarg])

super_function(50, -1, 1.56, 10, 20, 300, name="Peter", age=38)

sum =>  380.56
name => Peter
age => 38


## 5. Introduction to NumPy

NumPy is a Python library used to work with matrices. It also has functions for working with linear algebra or Fourier transforms. NumPy stands for Numerical Python.

<img src="Figures/NumPy_logo.png" alt="Drawing" style="width: 400px;"/>

As we have seen, in Python we already have lists that serve the purpose of working with arrays, but their processing is slow. NumPy aims to provide an array object that is up to 50 times faster than traditional Python lists, something important for data science where speed and resources are very important. In addition, they differ from lists in that NumPy arrays cannot be resized, all elements must be of the same type, and inter-array operations are allowed.

The NumPy source code can be found in this github repository https://github.com/numpy/numpy

The array object in NumPy is called ``ndarray``, it provides many support functions that make working with ``ndarray`` very easy. You can create an ndarray with NumPy with the ``.array()`` function.

To create an ndarray, we can use a list, tuple or any array object to the ``array()`` method, and it will become an ``ndarray``. You can create ``ndarrays`` of different dimensions. 0-D are just numbers, 1-D are vectors, 2-D are arrays, and so on.

The main attributes that can be used with ``ndarrays`` are:

| Attribute | Description | 
|-----------|-------------|
| ndim |number of dimensions of the matrix| 
| shape |number of elements of each dimension of the array|
| size |total number of elements in the matrix|
| dtype | type of the elements of the matrix|
| astype |to change the type of the elements of the matrix|

In [47]:
#Example

array0 = np.array([(0,1)])

array1 = np.array(([0],[1]))

array2 = np.array([0,1])

Lets print the arrays. Whats the difference between them?

In [48]:
print(array0)
print(array1)
print(array2)

[[0 1]]
[[0]
 [1]]
[0 1]


In [49]:
print(array0.ndim)
print(array0.shape)

print(array1.ndim)
print(array1.shape)

print(array2.ndim)
print(array2.shape)

2
(1, 2)
2
(2, 1)
1
(2,)


In [50]:
#Example of multi-dimension

array0 = np.array([(0,1),(2,3)])

array1 = np.array([(0,1,2),(3,4,5), (6,7,8)])

array2 = np.array([[(0,1,2), (3,4,5)], [(6,7,8), (9,10,11)]])

print(array0.ndim)
print(array0.shape)

print(array1.ndim)
print(array1.shape)

print(array2.ndim)
print(array2.shape)

2
(2, 2)
2
(3, 3)
3
(2, 2, 3)


### 5.1 Accessing values in an ndarray

An array element can be accessed by referring to its index number. Indexes in NumPy arrays start with 0, which means that the first element has index 0, and the second has index 1, etc.

To access elements of 2-D arrays we can use comma-separated integers representing the dimension and index of the element. For 3-D arrays or higher, you must add as many comma-separated integers as there are dimensions in the array.

In [51]:
#Example

arr = np.array([[1,2,3,4,5], [6,7,8,9,10]])

print(arr[0, 1])

2


Slicing can also be used in ``ndarrays``. In this case: ``array[indice_primer_elemento:indice_final:step]``

In [52]:
# Example

arr = np.array([[1, 2, 3, 4, 5], [6, 7, 8, 9, 10]])

# returns the index 2 of the two dimensions:

arr[:, 2]

# returns the interval between index 1 and 4 (not included) of the two dimensions:
arr[:, 1:4]

# return all elements in even positions:
arr[:, ::2]

array([[ 1,  3,  5],
       [ 6,  8, 10]])

<img src="Figures/numpy_indexing.png" alt="Drawing" style="width: 200px;"/>

####  <font color='red'>Try it yourself!</font> 

<font color='red'>Select the elements of each colored set of elements in the previous figure.</font> 

In [53]:
arr = np.array([[0,1,2,3,4,5],[10,11,12,13,14,15],[20,21,22,23,24,25],[30,31,32,33,34,35],[40,41,42,43,44,45],[50,51,52,53,54,55]])
arr

# New array with selection of red elements:
arr2 = arr[:,2]
print(arr2)

# New array with selection of orange elements:
arr3 = arr[0,3:5]
print(arr3)

# New array with selection of green elements:
arr4 = arr[2:5:2, ::2]
print(arr4)

# New array with selection of blue elements:
arr5 = arr[4:, 4:]
print(arr5)

[ 2 12 22 32 42 52]
[3 4]
[[20 22 24]
 [40 42 44]]
[[44 45]
 [54 55]]


### For more information about NumPy

* http://numpy.scipy.org
* http://scipy.org/Tentative_NumPy_Tutorial
* https://numpy.org/learn/

# 6. Introduction to Pandas

**Pandas is one of the most important libraries in Python for Data Science!!!!** Basically it will allow us to have ``Datasets`` opened as if they were an excel spreadsheet, that is, in matrix form with two dimensions and with indexed columns and indexes.

You can learn more in the documentation: https://pandas.pydata.org/docs/

<img src="Figures/Pandas.jpg" alt="Drawing" style="width: 800px;"/>


Pandas is used to:
* Operate on the whole dataset or on a row or column (vectors) with high computational efficiency
* Helps to clean data and fix missing data
* Access subsets of data
* Add or remove columns with new features.
* Group data by characteristics.
* Very efficient in joining data from different sources.
* Work with time series.

## 6.1 Read data from a csv

Most of the time we will load the files through a .csv file. Below you can see the basic code that will be used most of the time:

In [54]:
import pandas as pd
london = pd.read_csv('Data/block_13_daily.csv')
london

,LCLid,day,energy_median,energy_mean,energy_max,energy_count,energy_std,energy_sum,energy_min
0,MAC000113,2011-12-14,1.0910,1.065250,1.674,28,0.305397,29.827,0.262
1,MAC000113,2011-12-15,0.3710,0.546521,1.434,48,0.410120,26.233,0.119
2,MAC000113,2011-12-16,0.3205,0.534708,2.034,48,0.464431,25.666,0.129
3,MAC000113,2011-12-17,0.2570,0.421500,1.068,48,0.274387,20.232,0.115
4,MAC000113,2011-12-18,0.3985,0.565333,2.218,48,0.450712,27.136,0.134
...,...,...,...,...,...,...,...,...,...
32987,MAC005416,2014-02-24,0.0660,0.111437,0.809,48,0.126465,5.349,0.050
32988,MAC005416,2014-02-25,0.0645,0.072521,0.238,48,0.030756,3.481,0.051
32989,MAC005416,2014-02-26,0.0675,0.107917,0.343,48,0.080074,5.180,0.049
32990,MAC005416,2014-02-27,0.0910,0.104250,0.296,48,0.054844,5.004,0.047


## 6.2 Data inspection

If we want to know what dimensions this DataFrame has, we can look at its 'shape'.

In [55]:
london.shape

(32992, 9)

There are 32992 entries and 9 characteristics for each entry.

It can be very useful to view the first or last entries of the dataset to better understand this data:

In [56]:
london.head(10)

,LCLid,day,energy_median,energy_mean,energy_max,energy_count,energy_std,energy_sum,energy_min
0,MAC000113,2011-12-14,1.0910,1.065250,1.674,28,0.305397,29.827,0.262
1,MAC000113,2011-12-15,0.3710,0.546521,1.434,48,0.410120,26.233,0.119
2,MAC000113,2011-12-16,0.3205,0.534708,2.034,48,0.464431,25.666,0.129
3,MAC000113,2011-12-17,0.2570,0.421500,1.068,48,0.274387,20.232,0.115
4,MAC000113,2011-12-18,0.3985,0.565333,2.218,48,0.450712,27.136,0.134
5,MAC000113,2011-12-19,0.6980,0.746479,2.204,48,0.488694,35.831,0.131
6,MAC000113,2011-12-20,0.2485,0.436354,1.551,48,0.362781,20.945,0.139
7,MAC000113,2011-12-21,0.2050,0.426542,1.818,48,0.477511,20.474,0.119
8,MAC000113,2011-12-22,0.1830,0.181854,0.259,48,0.040321,8.729,0.119
9,MAC000113,2011-12-23,0.1860,0.185646,0.268,48,0.040754,8.911,0.121


In [57]:
london.tail()

,LCLid,day,energy_median,energy_mean,energy_max,energy_count,energy_std,energy_sum,energy_min
32987,MAC005416,2014-02-24,0.0660,0.111437,0.809,48,0.126465,5.349,0.050
32988,MAC005416,2014-02-25,0.0645,0.072521,0.238,48,0.030756,3.481,0.051
32989,MAC005416,2014-02-26,0.0675,0.107917,0.343,48,0.080074,5.180,0.049
32990,MAC005416,2014-02-27,0.0910,0.104250,0.296,48,0.054844,5.004,0.047
32991,MAC005416,2014-02-28,0.0630,0.063000,0.063,1,NaN,0.063,0.063


If we want to know the name of the columns or indexes, we can access them:

In [58]:
london.columns

Index(['LCLid', 'day', 'energy_median', 'energy_mean', 'energy_max',
       'energy_count', 'energy_std', 'energy_sum', 'energy_min'],
      dtype='object')

In [59]:
london.index

RangeIndex(start=0, stop=32992, step=1)

Finally, to have a very good summary, it is very useful to use the ``describe`` method.

In [60]:
london.describe()

,energy_median,energy_mean,energy_max,energy_count,energy_std,energy_sum,energy_min
count,32992.000000,32992.000000,32992.000000,32992.000000,32865.000000,32992.000000,32992.000000
mean,0.165632,0.223007,0.879596,47.776128,0.182819,10.660024,0.064419
std,0.158396,0.178484,0.672407,3.035524,0.147394,8.565058,0.076632
min,0.000000,0.000000,0.000000,1.000000,0.000000,0.000000,0.000000
25%,0.052000,0.085432,0.348000,48.000000,0.069162,4.046000,0.016000
50%,0.118500,0.176844,0.775000,48.000000,0.149607,8.458000,0.035000
75%,0.225000,0.311276,1.248000,48.000000,0.264400,14.910500,0.093000
max,1.608000,1.407458,6.897000,48.000000,1.104101,67.558000,1.068000


## 6.3 Select data

If we want to work or select only a part of the dataset, we can use **[ ]** after the dataset name, and indicate which columns we want to select.

In [61]:
london[['day', 'energy_max']].head()

,day,energy_max
0,2011-12-14,1.674
1,2011-12-15,1.434
2,2011-12-16,2.034
3,2011-12-17,1.068
4,2011-12-18,2.218


If, on the other hand, you want to select a part of the rows, it will be indicated using *slicing* as in the strings:

In [62]:
london.iloc[0]

LCLid             MAC000113
day              2011-12-14
energy_median         1.091
energy_mean         1.06525
energy_max            1.674
energy_count             28
energy_std         0.305397
energy_sum           29.827
energy_min            0.262
Name: 0, dtype: object

#### loc vs iloc

To select more specific parts (more than one column, several columns and only a few rows of these,...), indexing with loc or iloc will be used. The only difference between the two is that the first one is based on the *label* of the rows or columns, while iloc is based on the *integer* of these. With an example it is better understood:

In [63]:
london.loc[3:5,['day', 'energy_max']]

,day,energy_max
3,2011-12-17,1.068
4,2011-12-18,2.218
5,2011-12-19,2.204


In [64]:
london.iloc[3:5,1:3]

,day,energy_median
3,2011-12-17,0.2570
4,2011-12-18,0.3985


Note that the index will not always be a number and can be another parameter. For example, we can set the day as the index.

## 6.4 Adding/Removing Data

To add new columns/rows we can simply use `df[new_column]` or `df.loc[new_row]` and to remove we can use the `drop` method with the parameter `axis` being 0 for row and 1 for column. 

In [65]:
# Let's create a smaller copy of our dataset:

df = london.head(5).copy()
df

,LCLid,day,energy_median,energy_mean,energy_max,energy_count,energy_std,energy_sum,energy_min
0,MAC000113,2011-12-14,1.0910,1.065250,1.674,28,0.305397,29.827,0.262
1,MAC000113,2011-12-15,0.3710,0.546521,1.434,48,0.410120,26.233,0.119
2,MAC000113,2011-12-16,0.3205,0.534708,2.034,48,0.464431,25.666,0.129
3,MAC000113,2011-12-17,0.2570,0.421500,1.068,48,0.274387,20.232,0.115
4,MAC000113,2011-12-18,0.3985,0.565333,2.218,48,0.450712,27.136,0.134


In [66]:
# Creating a new column

df['type'] = 'industrial'

# Creating a new row
df.loc[5] = 0


df

,LCLid,day,energy_median,energy_mean,energy_max,energy_count,energy_std,energy_sum,energy_min,type
0,MAC000113,2011-12-14,1.0910,1.065250,1.674,28,0.305397,29.827,0.262,industrial
1,MAC000113,2011-12-15,0.3710,0.546521,1.434,48,0.410120,26.233,0.119,industrial
2,MAC000113,2011-12-16,0.3205,0.534708,2.034,48,0.464431,25.666,0.129,industrial
3,MAC000113,2011-12-17,0.2570,0.421500,1.068,48,0.274387,20.232,0.115,industrial
4,MAC000113,2011-12-18,0.3985,0.565333,2.218,48,0.450712,27.136,0.134,industrial
5,0,0,0.0000,0.000000,0.000,0,0.000000,0.000,0.000,0


In [67]:
# Changing the value of second row of the new column

df.loc[2, 'type'] = 'residential'
df

,LCLid,day,energy_median,energy_mean,energy_max,energy_count,energy_std,energy_sum,energy_min,type
0,MAC000113,2011-12-14,1.0910,1.065250,1.674,28,0.305397,29.827,0.262,industrial
1,MAC000113,2011-12-15,0.3710,0.546521,1.434,48,0.410120,26.233,0.119,industrial
2,MAC000113,2011-12-16,0.3205,0.534708,2.034,48,0.464431,25.666,0.129,residential
3,MAC000113,2011-12-17,0.2570,0.421500,1.068,48,0.274387,20.232,0.115,industrial
4,MAC000113,2011-12-18,0.3985,0.565333,2.218,48,0.450712,27.136,0.134,industrial
5,0,0,0.0000,0.000000,0.000,0,0.000000,0.000,0.000,0


In [68]:
# Removing the added column  axis=(0 for row, 1 for column)
df.drop('type', axis=1)

# removing the 5th row:
df.drop(5, axis=0)

,LCLid,day,energy_median,energy_mean,energy_max,energy_count,energy_std,energy_sum,energy_min,type
0,MAC000113,2011-12-14,1.0910,1.065250,1.674,28,0.305397,29.827,0.262,industrial
1,MAC000113,2011-12-15,0.3710,0.546521,1.434,48,0.410120,26.233,0.119,industrial
2,MAC000113,2011-12-16,0.3205,0.534708,2.034,48,0.464431,25.666,0.129,residential
3,MAC000113,2011-12-17,0.2570,0.421500,1.068,48,0.274387,20.232,0.115,industrial
4,MAC000113,2011-12-18,0.3985,0.565333,2.218,48,0.450712,27.136,0.134,industrial


## 6.5 Filtering data

Another way to select only a part of the dataset is to apply *Boolean indexing*, which is nothing more than applying a filter to the data. For example, we will select only those days in which the maximum energy has been greater than 1.

In [69]:
london['energy_max'] > 5

0        False
1        False
2        False
3        False
4        False
         ...  
32987    False
32988    False
32989    False
32990    False
32991    False
Name: energy_max, Length: 32992, dtype: bool

As you can see, now we have selected those columns that meet our condition. What if we put it directly inside the dataset when we select only one part?

In [70]:
# Here we see that all the rows are no longer displayed

london[london['energy_max'] > 5]

,LCLid,day,energy_median,energy_mean,energy_max,energy_count,energy_std,energy_sum,energy_min
2976,MAC002510,2012-10-28,0.2695,0.470500,5.087,48,0.991433,22.584,0.021
2986,MAC002510,2012-11-07,0.0510,0.264750,5.095,48,0.752222,12.708,0.039
2994,MAC002510,2012-11-15,0.0295,0.171500,5.253,48,0.753426,8.232,0.019
3003,MAC002510,2012-11-24,0.1570,0.465229,6.816,48,1.047584,22.331,0.020
3008,MAC002510,2012-11-29,0.0215,0.371896,6.357,48,0.994860,17.851,0.019
3010,MAC002510,2012-12-01,0.8920,0.871542,5.238,48,0.968662,41.834,0.020
3036,MAC002510,2012-12-27,0.0170,0.488750,6.036,48,1.065459,23.460,0.016
3046,MAC002510,2013-01-06,0.3665,0.700521,5.440,48,0.936179,33.625,0.084
3095,MAC002510,2013-02-24,0.1370,0.567167,6.172,48,0.961951,27.224,0.020
3108,MAC002510,2013-03-09,0.7800,0.752292,6.897,48,1.067885,36.110,0.020


###  <font color='red'>Try it yourself!</font> 

<font color='red'>Obtain a summary table of the parameters for the counter with ID 'MAC000113' using the method `describe`, then use `.isna().sum()` to see if there are any missing values.</font> 

In [71]:
# Obtain a DataFrame with only the values of the counter with ID 'MAC000113'
a = london[london['LCLid']=='MAC000113']


# Use the method .describe() to obtain the summary table of the parameters
display(a.describe())

# We see that there is some missing value in one of the columns. This can also be checked with the method .isna().sum()
display(a.isna().sum())

,energy_median,energy_mean,energy_max,energy_count,energy_std,energy_sum,energy_min
count,808.000000,808.000000,808.000000,808.000000,801.000000,808.000000,808.000000
mean,0.425918,0.519871,1.460573,47.553218,0.316776,24.656976,0.157256
std,0.180692,0.173581,0.558816,4.412963,0.125080,8.470674,0.079395
min,0.154000,0.154458,0.195000,1.000000,0.027048,0.195000,0.078000
25%,0.272375,0.403198,1.088000,48.000000,0.246581,19.162250,0.117000
50%,0.422750,0.525052,1.512000,48.000000,0.326376,25.076000,0.128000
75%,0.537000,0.627714,1.845250,48.000000,0.396441,30.043250,0.177000
max,1.091000,1.098521,3.338000,48.000000,0.916933,52.729000,1.068000


LCLid            0
day              0
energy_median    0
energy_mean      0
energy_max       0
energy_count     0
energy_std       7
energy_sum       0
energy_min       0
dtype: int64

## 6.6 Interesting basic functions

Here you can see a list of the most interesting functions to look at in the documentation, with some examples:

| Function | Description |
|-----------|---------------|
| count() |Number of non-null observations|  
| sum() |Sum of values|
| mean() |Average of observations | 
| median() |Median of observations |
| min() |Minimum value|
| max() |Maximum value|
| std() |Standard deviation|
| var() | Variance|
| value_counts() | Unique values in each column - freq table|
| nunique() | Number of distinct values|
|  isnull() | Mask for null or Nan|

Practice with some of these functions.

###  <font color='red'>Try it yourself!</font> 

* What is the total energy of counter 'MAC000113'? 
* How many different counters are there? 
* What was the maximum energy value on May 15, 2013?</font> 

In [87]:
# Total energy value for ID 'MAC000113'

id_MAC000113 = london[london['LCLid']=='MAC000113']
print(id_MAC000113['energy_count'].sum())

# How many different counters are there and what are their names?

print(london['LCLid'].value_counts())

# What was the maximum daily energy consumption value on May 15, 2013?

print(london[london['day']=='2013-05-15']['energy_max'].max())

38423
LCLid
MAC000113    808
MAC005269    750
MAC005270    750
MAC005313    744
MAC005317    744
MAC005319    744
MAC005367    737
MAC005392    733
MAC005400    733
MAC005404    733
MAC005405    733
MAC005403    733
MAC005411    732
MAC005408    732
MAC005407    732
MAC005412    732
MAC005415    731
MAC005416    731
MAC000301    728
MAC000309    725
MAC005331    703
MAC002662    666
MAC002663    666
MAC002676    666
MAC002666    665
MAC002702    665
MAC002757    660
MAC002750    660
MAC002764    660
MAC002761    660
MAC002754    660
MAC002756    660
MAC002509    602
MAC002512    602
MAC002519    600
MAC002518    600
MAC002514    600
MAC002558    600
MAC002530    600
MAC002525    600
MAC002522    600
MAC002539    600
MAC002513    600
MAC002517    600
MAC002605    599
MAC002557    598
MAC005322    531
MAC002510    501
MAC002767    497
MAC005274    286
Name: count, dtype: int64
2.7160001


## 6.7 Sort

If you want to sort values, you can use ``sort_values``. With an example it's better understood:


In [89]:
london.sort_values(by='energy_max', ascending= False, inplace=True)
london.head(10)

,LCLid,day,energy_median,energy_mean,energy_max,energy_count,energy_std,energy_sum,energy_min
3108,MAC002510,2013-03-09,0.7800,0.752292,6.897,48,1.067885,36.110,0.020
3003,MAC002510,2012-11-24,0.1570,0.465229,6.816,48,1.047584,22.331,0.020
3123,MAC002510,2013-03-24,0.9325,0.933542,6.428,48,1.104101,44.810,0.023
3008,MAC002510,2012-11-29,0.0215,0.371896,6.357,48,0.994860,17.851,0.019
3236,MAC002510,2013-07-15,0.0505,0.262250,6.297,48,0.918086,12.588,0.020
3125,MAC002510,2013-03-26,0.0295,0.520958,6.295,48,1.050477,25.006,0.020
3095,MAC002510,2013-02-24,0.1370,0.567167,6.172,48,0.961951,27.224,0.020
3109,MAC002510,2013-03-10,0.3105,0.659375,6.147,48,1.008473,31.650,0.021
3277,MAC002510,2013-08-25,0.0500,0.239250,6.070,48,0.881812,11.484,0.021
3036,MAC002510,2012-12-27,0.0170,0.488750,6.036,48,1.065459,23.460,0.016


## 6.8 Group

Finally, a very useful function is the group function (group_by).

For example, imagine that you want to group all the data by counter number and take some parameter from this counter (maximum, minimum, etc.).

The result of this operation will be another dataframe with the data grouped according to a variable. In addition, you must also indicate which *aggregation* function you want to use (max, min, mean, count, sum, etc.).

Let's see an example!

In [90]:
smart_meters = london[['LCLid','energy_max', 'energy_min', 'energy_sum']].groupby('LCLid').count()
smart_meters

,energy_max,energy_min,energy_sum
LCLid,,,
MAC000113,808,808,808
MAC000301,728,728,728
MAC000309,725,725,725
MAC002509,602,602,602
MAC002510,501,501,501
MAC002512,602,602,602
MAC002513,600,600,600
MAC002514,600,600,600
MAC002517,600,600,600


# 7. Exercise 1

Using the dataset provided, write down a function that receives an input from the user with the LCLid of a counter and a year. Your function must return a `dataframe` containing all the data from that counter in that year.

Also, print all columns of the line corresponding to the first letter of your surname based on the alphabetic order (A = 1st line, B = 2nd line ...)

In [91]:
london.head()

,LCLid,day,energy_median,energy_mean,energy_max,energy_count,energy_std,energy_sum,energy_min
3108,MAC002510,2013-03-09,0.7800,0.752292,6.897,48,1.067885,36.110,0.020
3003,MAC002510,2012-11-24,0.1570,0.465229,6.816,48,1.047584,22.331,0.020
3123,MAC002510,2013-03-24,0.9325,0.933542,6.428,48,1.104101,44.810,0.023
3008,MAC002510,2012-11-29,0.0215,0.371896,6.357,48,0.994860,17.851,0.019
3236,MAC002510,2013-07-15,0.0505,0.262250,6.297,48,0.918086,12.588,0.020


In [105]:
lclid = input('Choose an id: ')
year = input('Choose an year: ')


def mifuncion (lclid=None, year=None) -> pd.DataFrame: 
    if not lclid or not year:
        print("#ERROR# Introduce todos los valores")
        return
    
    df = london.copy()  
    df['año'] = df['day'].str[:4].astype(int)

    df = df[df['LCLid'] == lclid]
    df = df[df['año'] == int(year)]

    return(df)


# My surname is Mañé
resultado = mifuncion(lclid,year)
display(resultado.columns)
print(resultado.iloc[11])


Index(['LCLid', 'day', 'energy_median', 'energy_mean', 'energy_max',
       'energy_count', 'energy_std', 'energy_sum', 'energy_min', 'año'],
      dtype='object')

LCLid             MAC002510
day              2013-04-21
energy_median        0.3895
energy_mean        0.641354
energy_max            5.606
energy_count             48
energy_std         0.932376
energy_sum           30.785
energy_min            0.022
año                    2013
Name: 3151, dtype: object
